### PR TITLE
port: turn state set locale

### LINKED ISSUE #3188: port: Make sure "turn.locale" is correctly set in TurnContext, enabling locale can be changed during runtime (#5019)
The changes in [Make sure "turn.locale" is correctly set in TurnContext, enabling locale can be changed during runtime (#5019)](https://github.com/microsoft/botbuilder-dotnet/pull/5019) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Closes: #4983 #3993 
In this PR, fix a bug of set locale in TurnState. The locale info will be set to the value of key "turn".
User can set locale by SetProperty action.
For example, suppose the original locale is "en-US":
```
{
    "$schema": "../../tests.schema",
    "$kind": "Microsoft.AdaptiveDialog",
    "generator": "test.lg",
    "triggers": [
        {
            "$kind": "Microsoft.OnBeginDialog",
            "actions": [
                {
                    "$kind": "Microsoft.SetProperty",
                    "property": "turn.locale",
                    "value": "de-DE"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${string(1.122)}"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${formatNumber(1.12345, 4)}"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${addDays('2018-01-02T02:00:00.000Z', 4, 'D')}"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${formatNumber(3.1415926, 5)}"
                }
            ]
        }
    ]
}
```
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by joshgummersall**
Fixes #3188

### CODE REVIEW

**Comment by cosmicshuai** [comment]
Hi @joshgummersall 
I think we should fix the gets and sets of locale in turnContext. This ensures the locale can be set correctly as setValueToPath.
```
    /**
     * Gets the locale stored in the turnState.
     */
    public get locale(): string {
        return this._turnState.get(this._turnLocale);
    }

    /**
     * Sets the locale stored in the turnState.
     */
    public set locale(value: string) {
        this._turnState.set(this._turnLocale, value);
    }

```

**Comment by joshgummersall** [comment]
The problem is the symbol that is used to store turn state is intentionally not exported from the dialogs package. I left a more detailed note on the original issue.

**Comment by chon219** [comment]
Can we move the symbol from `botbuilder-dialogs` to `botbuilder-core` and let `botbuilder-dialogs` import it?
Without this symbol, `TurnContext.locale` cannot set the locale value to the right place, as a result, the memory object would look like { 'turn.locale': 'en-us' } not the expected one { 'turn': { 'locale': 'en-us' } }.

**Comment by joshgummersall** [comment]
It seems like we don't want to expose the symbol as a way of limiting access to that piece of state. Why can't we just use the turn.state locale if it exists and defer to the context locale if it doesn't exist?

**Comment by chon219** [comment]
Can we just use `turn` instead of symbol as the key to the turn memory? We may also need to modify the get/set methods in turnContext to the following:
```
const turnMemory = this._turnState.get('turn');
turnMemory.locale = value;
```

**Comment by joshgummersall** [comment]
To me it's more about a separation of concerns. The current turn state object is defined in the dialogs package. The TurnContext object is a core component that exists independent of the dialogs package. Why are we conflating these packages and spreading the locale state across them?

File: libraries/botbuilder-dialogs-adaptive-testing/tests/resources/LGGeneratorTests/ManuallySetLocale.test.dialog [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{
```
Comment (stevengum) [inline libraries/botbuilder-dialogs-adaptive-testing/tests/resources/LGGeneratorTests/ManuallySetLocale.test.dialog:1]: This file is different from the corresponding [.NET `manuallySetLocale.dialog` file](https://github.com/microsoft/botbuilder-dotnet/blob/00d1155d89186d3d6378d5df2fee099e00ed0a4f/tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests/Fallback/manuallySetLocale.dialog), is this okay?

**Comment by cosmicshuai** [comment]
Hi @joshgummersall,
I have a fix for the locale stuff. See:https://github.com/microsoft/botbuilder-dotnet/pull/5156
Could we wait for a while. After the dotnet SDK PR is merged, I will port the fix to JS right away.

**Comment by joshgummersall** [comment]
@cosmicshuai sounds good.

**Pull request closed without merging**